monitor/monitor.go

```diff
@@ -46,45 +43,33 @@ type Monitor struct {
 
 // Run starts monitoring.
 func (m *Monitor) Run(npages int) {
-	eventConfig.NumPages = npages
-	t := time.NewTicker(5 * time.Second)
-	emptyMetaBuf = generateHealthCheckMsg()
-	for range t.C {
-		// This loop is meant to loop forever so our process does not become
-		// defunct and we still print the stat for connected monitors.
-		if monitorEvents != nil {
-			m.dumpStat()
-		}
-	}
-}
-
-func purgeInactiveListeners() {
-	var next *list.Element
-	for e := listeners.Front(); e != nil; e = next {
-		client := e.Value.(net.Conn)
-		next = e.Next()
-		if _, err := client.Write(emptyMetaBuf); err != nil {
-			log.WithError(err).Warn("metadata write failed; removing client")
-			client.Close()
-			listeners.Remove(e)
-			continue
-		}
-	}
-}
+	log.Info("Starting monitoring traffic")
+	c := bpf.DefaultPerfEventConfig()
+	c.NumPages = npages
 
-func generateHealthCheckMsg() []byte {
-	meta := &payload.Meta{Size: uint32(0)}
-	metaBuf, err := meta.MarshalBinary()
+	me, err := bpf.NewPerCpuEvents(c)
 	if err != nil {
-		log.WithError(err).Info("connection probe")
+		log.WithError(err).Error("Error while starting monitor")
+		return
 	}
-	return metaBuf
-}
+	monitorEvents = me
 
-func numListeners() int {
-	mutex.Lock()
-	defer mutex.Unlock()
-	return listeners.Len()
+	// Main event loop
+	for {
+		todo, err := monitorEvents.Poll(pollTimeout)
+		if err != nil {
+			log.WithError(err).Error("Error in Poll")
+			if err == syscall.EBADF {
+				break
+			}
+		}
+		if todo > 0 {
+			if err := monitorEvents.ReadAll(m.receiveEvent, m.lostEvent); err != nil {
+				log.WithError(err).Warn("Error received while reading from perf buffer")
+			}
+		}
+		m.dumpStat()
```

Comment:
This will print the stats on every loop and that could be as often as every packet, right? Could we rate limit it by storing the time when we write it and checking that enough time has passed?